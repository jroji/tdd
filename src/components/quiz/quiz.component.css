.next {
  background: var(--color-orange);
  padding: var(--spacing);
  margin-top: var(--spacing);
  border-radius: var(--answer-border-radius);
  border: none;
  width: 100px;
  align-self: flex-end;
  font-size: 15px;
  color: var(--color-white);
  order: 1;
  cursor: pointer;
}

#question {
  font-size: 1.5em;
  color: var(--color-text);
  margin-top: calc(var(--spacing) * 4);
}

.hidden {
  display: none;
}