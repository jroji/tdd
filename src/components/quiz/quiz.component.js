import { AnswersComponent } from "../answers/answers.component";

/**
 * Crea un Quiz con preguntas y respuestas sobre países
* @param {HTMLElement} htmlContainer
* @param {countryService} countryService
 */
export class Quiz {

  constructor(htmlContainer, countryService) {
    this.htmlContainer = htmlContainer;
    this.countryService = countryService;
    this.renderNextButton(htmlContainer);
    this.htmlContainer.addEventListener('answerClick', (event) => this.handleAnswer(event));
  }

  /**
   * Realiza una petición al servicio de países y renderiza las respuestas
   */
  async renderQuestion() {
    const { countryName, answers, correctAnswer } = await this.countryService.getQuestion();
    this.renderTitle(this.htmlContainer, countryName);
    if (!this.answersContainer) {
      this.answersContainer = this.renderAnswerContainer(this.htmlContainer);
    }
    this.answersContainer.innerHTML = '';
    this.nextButton.style.display = 'none';
    this.answers = new AnswersComponent(this.answersContainer, answers, correctAnswer)
  }

  /**
   * Pinta el botón de next y asocia el evento
   * @param {HTMLElement} htmlContainer
   */
  renderNextButton(htmlContainer) {
    const nextButton = this.createElement('button', 'Next', 'next-button');
    nextButton.className = 'next';
    nextButton.addEventListener('click', (event) => { this.renderQuestion(); });

    this.nextButton = nextButton;
    htmlContainer.appendChild(nextButton);
  }

  /**
   * Pinta una pregunta con el nombre del país a adivinar
   * @param {HTMLElement} htmlContainer
   * @param {string} countryName
   */
  renderTitle(htmlContainer, countryName) {
    let question = htmlContainer.querySelector('#question');
    if (question) {
      return question.textContent = `${countryName} is the capital of `;
    }

    question = this.createElement('h2', `${countryName} is the capital of `, 'question');
    question.id = 'question';

    htmlContainer.appendChild(question);
  }

  /**
   * Crea el contenedor para añadir las respuestas
   * @param {HTMLElement} htmlContainer
   */
  renderAnswerContainer(htmlContainer) {
    const container = this.createElement('section', '', 'answers-section');
    htmlContainer.appendChild(container);
    return container;
  }

  /**
   * Crea y devuelve un elemento HTML con la configuración proporcionada
   * @param {string} tag Etiqueta a crear
   * @param {string} content Contenido que se mostrará al usuario
   * @param {string} testId Identificador de test
   */
  createElement(tag, content, testId) {
    const element = document.createElement(tag);
    element.textContent = content;
    element.setAttribute('data-testid', testId);
    return element;
  }

  handleAnswer(event) {
    this.nextButton.style.display = 'block';
  }
}