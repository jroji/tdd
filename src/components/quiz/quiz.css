.next {
  background: var(--color-orange);
  padding: var(--spacing);
  border-radius: var(--answer-border-radius);
  border: none;
  width: 100px;
  align-self: flex-end;
  font-size: 15px;
  color: var(--color-white);
  order: 1;
}

#question {
  font-size: 1.5em;
}

.hidden {
  display: none;
}