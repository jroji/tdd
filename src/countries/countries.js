
const BASE_API = 'https://restcountries.eu/rest/v2/';

export class CountriesService {

  async getCountries() {
    const response = await fetch(`${BASE_API}all`);
    const countries = await response.json();
    return countries;
  }

  async getQuestion() {
    if (!this.countries) {
      this.countries = await this.getCountries();
    }

    const countryToGuess = this.getRandomCountry(this.countries);

    return {
      countryName: countryToGuess.name,
      answers: this.getAnswers(this.countries, countryToGuess),
      correctAnswer: countryToGuess.capital
    }
  }

  getRandomCountry(countries) {
    const max = this.countries.length;
    const idx = Math.floor(Math.random()*(max-0)+0);
    return countries[idx];
  }

  getAnswers(countries, countryToGuess) {
    return [
      countryToGuess.capital,
      this.getRandomCountry(countries).capital,
      this.getRandomCountry(countries).capital,
      this.getRandomCountry(countries).capital,
    ].sort(() => Math.random() - 0.5);
  }
}