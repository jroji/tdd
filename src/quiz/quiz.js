import { AnswersComponent } from "../answers/answers";

export class Quiz {
  constructor(htmlContainer, countryService) {
    this.htmlContainer = htmlContainer;
    this.countryService = countryService;
    this.renderNextButton(htmlContainer);
  }

  async renderQuestion() {
    const { countryName, answers, correctAnswer } = await this.countryService.getQuestion();
    this.renderTitle(this.htmlContainer, countryName);
    if (!this.answersContainer) {
      this.answersContainer = this.renderAnswerContainer(this.htmlContainer);
    }
    this.answersContainer.innerHTML = '';
    this.answers = new AnswersComponent(this.answersContainer, answers, correctAnswer)
  }

  renderNextButton(htmlContainer) {
    const nextButton = document.createElement('button');
    nextButton.setAttribute('data-testid', 'next-button');
    nextButton.className = 'next';
    nextButton.textContent = 'Next';
    nextButton.addEventListener('click', (event) => { this.renderQuestion(); });
    htmlContainer.appendChild(nextButton);
  }

  renderTitle(htmlContainer, countryName) {
    const question = htmlContainer.querySelector('#question');
    if (question) {
      return question.textContent = `${countryName} is the capital of `;
    }

    const nextButton = document.createElement('h2');
    nextButton.id = 'question';
    nextButton.setAttribute('data-testid', 'question');
    nextButton.textContent = `${countryName} is the capital of `;

    htmlContainer.appendChild(nextButton);
  }

  renderAnswerContainer(htmlContainer) {
    const container = document.createElement('section');
    htmlContainer.appendChild(container);
    return container;
  }
}