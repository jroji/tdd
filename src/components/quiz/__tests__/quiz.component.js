import { getAllByRole, getByTestId, getByText, waitFor } from "@testing-library/dom";
import { Quiz } from "../quiz.component";
import userEvent from "@testing-library/user-event";
import { CountriesService } from "../../countries/countries.service";
import { QuestionMock } from "../../countries/__mocks__/countries.service.mock";

// Gracias al automock de Jest, podemos hacer esto para mockear todas las respuestas
jest.mock("../../countries/countries.service");

// Para evitar que la petición devuelva undefined, mockeamos la respuesta del método que nos interesa
const mockCountriesAnswers = () => {
  CountriesService.prototype.getQuestion = jest.fn().mockReturnValue(QuestionMock);
};

describe('rendering quiz', () => {
  let dom;
  let countriesService;

  beforeAll(() => {
    mockCountriesAnswers();
    dom = document.createElement('div');
    countriesService = new CountriesService();

    new Quiz(dom, countriesService);
  });

  it ('renders correctly', () => {
    expect(dom).toMatchSnapshot();
  });

  beforeEach(() => {
    // Limpiamos los mocks para que en cada test, el numero de peticiones iniciales sea 0
    CountriesService.mockClear();
  });

  it('should render the next button', () => {
    expect(getByText(dom, 'Next')).toBeVisible();
  });

  it('next button shouldnt be visible without an answer', () => {
    expect(getByText(dom, 'Next'))
  });

  describe('clicking on next', () => {
    beforeEach(async () => {
      userEvent.click(getByText(dom, 'Next'));
      await waitFor(() => expect(getByTestId(dom, 'question')).toBeDefined());
    })

    it('should hide next button', () => {
      expect(getByText(dom, 'Next')).not.toBeVisible();
    });

    it('should render a question', () => {
      const title = getByTestId(dom, 'question');
      expect(title.textContent).toBe('Spain is the capital of ');
    });

    // Este test, sería probar funcionalidades internas, lo cual no nos interesa con el enfoque 'user-driven-testing'
    // it('next button should update the question', () => {
    //   expect(countriesService.getQuestion).toHaveBeenCalled();
    // })

    describe('after click on an answer', () => {

      beforeEach(() => {
        const answerSection = getByTestId(dom, 'answers-section');
        userEvent.click(getAllByRole(answerSection, 'button')[0]);
      });

      it('should render the next button', async () => {
        expect(getByText(dom, 'Next')).toBeVisible();
      });
    })

  })
});