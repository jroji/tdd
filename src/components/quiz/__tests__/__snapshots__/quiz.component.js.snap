// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering quiz renders correctly 1`] = `
<div>
  <button
    class="next"
    data-testid="next-button"
  >
    Next
  </button>
</div>
`;
